distributed/helpers.py

```diff
@@ -1,23 +1,20 @@
-
 import torch
 import torch.distributed as dist
 from utils import comm
 
-
-def get_memory_format(tensor):
-    """Helper routine to get the memory format"""
-    if tensor.is_contiguous(memory_format=torch.channels_last):
-        return torch.channels_last
-    else:
-        return torch.contiguous_format
-
-def sync_params(model):
+def init_params_for_shared_weights(model):
     """Helper routine to ensure shared weights are the same after initialization"""
     with torch.no_grad():
         # distributed sync step
         for param in model.parameters():
             if not hasattr(param, "is_shared_mp"):
-                param.is_shared_mp = ["model"]
+                # all sharded weights manually annotate this field
+                # if weight doesnt have annotation, then it is a shared weight
+                # layers like patch-embed, decoder head, pos-embed are fully
+                # shared (and not sharded) in this example
+                param.is_shared_mp = ["tp-cp"] # only TP-CP implemented for now
+                # careful about this stuff..
+                param.mark_for_reduction = [] # not all params need special handling
```

Comment:
I think the mark for reduction is not necessary, why not use is_shared_mp for that purpose. What we have in Makani is the following: we annotate each tensor what its sharing is, f the tensor has no is_shared_mp then we default to some behavior (In this case not do anything special with it). Parameters is per module parameter, so you can have both inside the same module, it is no problem.

Reply:
I wanted to distinguish between shared params (shared across some model parallel group) which need the same init, and shared params that need an additional allreduce due to the input activation gradient being sharded across the same model parallel group. These two sets of parameters are unequal I think (some shared params do not need the additional allreduce; example, if you look at the decoder head, a linear layer, it shares it parameters across the full model parallel group but the input activation gradient is also shared - so the wgrads will be in sync. The only allreduce they need is across the data parallel group where the input activation gradient is also shared (each model parallel group gets a different batch) and that allreduce is the normal DDP allreduce (but averaged due to the MSE loss). Do you feel this is correct?

Reply:
I think it is right

Reply:
The question is, is there a simple way to identify whether the grads need reductions or not? I guess not without some user intervention, or? That would be good to think about.